mmaction/datasets/pipelines/augmentations.py

```diff
@@ -868,6 +870,406 @@ def __repr__(self):
+        width = x2 - x1
+        height = y2 - y1
+
+        for label_index in gt_bboxes:
+            for tube in gt_bboxes[label_index]:
+                tube -= np.array([[x1, y1, x1, y1]], dtype=np.float32)
+
+                x_center = 0.5 * (tube[:, 0] + tube[:, 2])
+                y_center = 0.5 * (tube[:, 1] + tube[:, 3])
+
+                if np.any([
+                        x_center < 0, y_center < 0, x_center > width,
+                        y_center > height
+                ]):
+                    continue
+
+                # clip box
+                tube[:, 0] = np.maximum(0, tube[:, 0])
+                tube[:, 1] = np.maximum(0, tube[:, 1])
+                tube[:, 2] = np.minimum(width, tube[:, 2])
+                tube[:, 3] = np.minimum(height, tube[:, 3])
+
+                out_tubes[label_index].append(tube)
+
+        results['imgs'] = imgs
+        results['img_shape'] = imgs[0].shape[:2]
+        results['gt_bboxes'] = out_tubes
+        return results
+
+    def __repr__(self):
+        repr_str = (f'{self.__class__.__name__}('
+                    f'cuboid_settings={self.cuboid_settings})')
+        return repr_str
+
+
+@PIPELINES.register_module()
+class TubePad:
+    """Pad mean values into the input images with a expanding ratio for output
+    images in a certain probability.
+
+    Required keys are "imgs", "gt_bboxes", "img_shape", added or modified keys
+    are "pad", "imgs", "img_shape", "gt_bboxes".
+
+    Args:
+        expand_ratio (float): Probability of implementing tube padding.
+            Default: 0.5.
+        max_expand_ratio (float): Max expand ratio indicating expand ratio
+            range ([1, ``max_expand_ratio``]) to randomly sample from.
+            Default: 4.0.
+        mean_values (Sequence[float] | None): Mean values of different
+            channels. Default: None.
+    """
+
+    def __init__(self,
+                 expand_ratio=0.5,
+                 max_expand_ratio=4.0,
+                 mean_values=None):
+        self.expand_ratio = expand_ratio
+        self.max_expand_ratio = max_expand_ratio
+        if mean_values is not None:
+            if not isinstance(mean_values, Sequence):
+                raise TypeError(
+                    'Mean must be list, tuple or np.ndarray, but got '
+                    f'{type(mean_values)}')
+            self.mean_values = np.array(mean_values)
+        else:
+            self.mean_values = None
+
+    def __call__(self, results):
+        imgs = results['imgs']
+        gt_bboxes = results['gt_bboxes']
+        img_h, img_w = results['img_shape']
+
+        pad = False
+        if np.random.rand() < self.expand_ratio:
+            pad = True
+
+        results['pad'] = pad
+
+        if pad:
+            expand_scale = np.random.uniform(1, self.max_expand_ratio)
+            h = int(img_h * expand_scale)
+            w = int(img_w * expand_scale)
+            outs = [
+                np.zeros((h, w, 3), dtype=np.float32)
```

Comment:
does float32 actually needed? how about uint8?

Reply:
Not sure, may be we should leave it for model check. And add todo here.